Making the index friendly to a switch of active index segments

vgkholla opened the pull request:
Compaction needs to atomically change the valid index segments after the completion of a cycle. `PersistentIndex` was built with the assumption that the reference to the valid index segments will not change. This commit removes that assumption and makes all functions in the index work with a view of the valid index segments at a certain point in time.

Built on OSx and Linux. Formatted.

**Reviewers: @pnarayanan, @nsivabalan**

| Class | Class, % | Method, % | Line, % | Test class |
| --- | --- | --- | --- | --- |
| `PersistentIndex` | 100% (7/ 7) | 98.2% (54/ 55) | 91.9% (513/ 558) | `IndexTest` |

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -625,16 +694,18 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
         // check journal
         List<JournalEntry> entries = journal.getEntriesSince(offsetToStart, storeToken.getInclusive());
         logger.trace("Journal based token, Time used to get entries: {}", (time.milliseconds() - startTimeInMs));
+        ConcurrentSkipListMap<Offset, IndexSegment> indexes = indexesRef.get();
```
can you confirm something. We take the snapshot of the index right here. There are chances that indexes change after this and the following lines are going to consider the snapshot only. We are still sure we don't miss out on any entries, bcoz new entries will definitely be part of journal and doesn't affect the correctness. Hope you also have thought about it. Can you confirm my understanding and there are no issues

nsivabalan replied:
minor. you can get the snapshot just before "if else block". Looks like you need it in both cases.

vgkholla replied:
There aren't any issues. For your second comment, the snapshot is taken in the right places (just after looking at the journal).

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -755,17 +828,16 @@ private StoreFindToken resetTokenIfRequired(StoreFindToken storeToken) {
 
   private long getTotalBytesRead(StoreFindToken newToken, List<MessageInfo> messageEntries,
-      Offset logEndOffsetBeforeFind) {
+      Offset logEndOffsetBeforeFind, ConcurrentSkipListMap<Offset, IndexSegment> indexes) {
```
can you add java docs

vgkholla replied:
Added.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -306,22 +314,72 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
+   * in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for a provided segment to add.
+   */
+  void changeIndexSegments(List<File> filesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
```
also, what is the expectation from persistence front, for these index files? Are they mapped already. Just wanted to ensure that our IndexPersistor doesn't need any fixes to accommodate this change.

vgkholla replied:
Renamed to `segmentFilesToAdd`.

vgkholla replied:
I think there isn't any requirement from the perspective of the index that the index segment files being added or removed be next to each other. 

For your second question, I'm not sure what you mean by "mapped already". There are files - the mapping is done during the `IndexSegment` instantiation. The only expectation is that none of the segments added or removed have offsets in the journal.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -869,10 +954,13 @@ private StoreFindToken findEntriesFromSegmentStartOffset(Offset initialSegmentSt
+              break;
+            }
             currentSegment = indexes.get(nextSegmentStartOffset);
+            // stop if ineligible because of last modified time
```
since you are at it, can you also fix line 969 to send a fileSpan as an arg

vgkholla replied:
Done.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1068,7 +1179,7 @@ FindInfo findDeletedEntriesSince(FindToken token, long maxTotalSizeOfEntries, lo
         // Get entries starting from the token Key in this index.
         newToken = findEntriesFromSegmentStartOffset(storeToken.getOffset(), storeToken.getStoreKey(), messageEntries,
-            new FindEntriesCondition(maxTotalSizeOfEntries, endTimeSeconds));
+            new FindEntriesCondition(maxTotalSizeOfEntries, endTimeSeconds), indexesRef.get());
```
minor. You could get the snapshot of indexes before "if else block" so that you can use it in either of both.

vgkholla replied:
I take the snapshot in the right place (right after getting entries from the journal).

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -306,22 +314,72 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
+   * {@link IndexSegment} instances.
+   * @param filesToAdd the backing files of the {@link IndexSegment} instances to add.
+   * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added has an offset higher than that
```
Good point. Check added.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -306,22 +314,72 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
+   * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added has an offset higher than that
+   * in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for a provided segment to add.
```
Changed.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -306,22 +314,72 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
+      segmentsToAdd.put(indexSegment.getStartOffset(), indexSegment);
+    }
+
+    indexesChangeLock.lock();
```
Good point Refactored this a bit.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -306,22 +314,72 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
+    indexesChangeLock.lock();
+    try {
+      if (segmentsToAdd.size() > 0) {
+        Offset lastSegmentToAddEndOffset = segmentsToAdd.lastEntry().getValue().getEndOffset();
```
I'm not adding from the last - just checking whether the end offset of the last entry >= journal's first offset since this is a tree map and it is sorted. Anyway, this has gone away in the refactoring.

vgkholla commented:
@nsivabalan Addressed your comments

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -339,37 +339,36 @@ private void setEndOffsets() throws IOException {
       for (Map.Entry<Offset, IndexSegment> entry : oldIndexSegments.entrySet()) {
         if (!segmentsToRemove.contains(entry.getKey())) {
           newIndexSegments.put(entry.getKey(), entry.getValue());
+        } else if (entry.getKey().compareTo(journal.getFirstOffset()) >= 0) {
```
1. The last index segment can have the start offset that lies within the journal right? Why is this check required. 2. This was bit confusing. Can you have two diff loops. One loop to go over segmentsToRemove and throw if any of the start offsets lie within journal. 2nd loop to create the newIndexSegments. bcoz, if you want to, you can combine the previous check (segmentFilesToAdd's ) in here itself. But I feel having it separately makes it more readable.

vgkholla replied:
Not sure what you mean in point 1 - the check is to make sure that `segmentsToRemove` doesn't contain anything that is past the journal start offset. The `if` checks to see if the old segment the iterator is on is present in `segmentsToRemove`. If it is not, all good. If it is, then the `else if` makes sure it is allowed to be there (i.e. not past the start offset of the journal). Were you referring to something else?

The bug I see here is that I should be comparing the end offset of the index segment to the journal's start offset (not the start offset). I'll fix that.

vgkholla replied:
Moved the checking to a `for` before the lock.

nsivabalan commented on ambry-store/src/test/java/com.github.ambry.store/IndexTest.java:
```diff
@@ -1793,4 +1864,26 @@ private void waitUntilExpectedState(Thread.State expectedState, long timeToCheck
+  private void partitionIndexSegments(Map<Offset, IndexSegment> toRemove, Set<Offset> toRetain) {
+    int i = 0;
+    for (Map.Entry<Offset, IndexSegment> indexSegmentEntry : state.index.getIndexSegments().entrySet()) {
+      if (indexSegmentEntry.equals(state.index.getIndexSegments().lastEntry()) || i % 3 != 0) {
```
you could do i++ here and remove the line 1886

vgkholla replied:
Would prefer to keep it this way for readability.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -306,22 +314,71 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
+      ConcurrentSkipListMap<Offset, IndexSegment> oldIndexSegments = indexesRef.get();
+      ConcurrentSkipListMap<Offset, IndexSegment> newIndexSegments = new ConcurrentSkipListMap<>();
+      for (Map.Entry<Offset, IndexSegment> entry : oldIndexSegments.entrySet()) {
+        if (!segmentsToRemove.contains(entry.getKey())) {
```
I don't see why a lock is needed. Doesn't the atomic reference suffice? Couldn't this method simply do the following:
- do the checks for whether any of the entries (in either `segmentsToAdd` or `segmentsToRemove`) is in the  trailing (journal-covered) part of the Index, if so throw. If not, it cannot in the future either since the trailing part only moves forward.
- Make the switch atomically as follows (there are some optimizations below that you can use even if you do need the lock):
```java
newIndexSegments = new ConcurrentSkipListMap<>(oldIndexSegments); // construct from the old.
newIndexSegments.putAll(segmentsToAdd);
newIndexSegments.keySet().removeAll(segmentsToRemove);
indexesRef.set(newIndexSegments);
```
I don't see a need for an `indexesChangeLock` in this flow and `addToIndex()` does not need it either (see my comment there). I get it if you think the changes that really need this is coming later on in compaction. Would like to hear anyway as to why the atomic reference change would not suffice.

vgkholla replied:
This is scenario I think the lock is required for. Consider two threads, one of them executing `addToIndex()` and the other `changeIndexSegments()`.

The one executing `changeIndexSegments()` has just executed the line copying all the existing index segments over to what will be the new map (`newIndexSegments = new ConcurrentSkipListMap<>(oldIndexSegments)`) but hasn't switched the atomic ref to the new map (i.e. `indexesRef.set(newIndexSegments)` has not been executed). The one executing `addToIndex()` decides that a rollover is required and adds an `IndexSegment` - note that this all to the "old" index segment map since the ref hasn't been switched yet. Now, when the thread executing `changeIndexSegments()` does the `set()` operation, the newly added `IndexSegment` is lost.

Let me know if you see any holes in the reasoning or still think a lock isn't required.

Took your suggestions on the refactoring. Makes the code clearer, thanks !

pnarayanan replied:
Ah, rollovers - did not account for that. Perhaps you need the lock after all. Will think about this.

pnarayanan replied:
I think it should be possible to get rid of the lock if this is the only reason to have it. It is a bit silly to actually have to lock a data structure by operations that are acting on mutually exclusive regions of it. I think at the very least, it should be possible to keep internal references to both the old and the new `indexRefs` and during the interim period where `changeIndexSegments()` is active, `addToIndex()` could add to both (and you can work out the ordering of variable updates for correctness). Something as follows:

```java
changeIndexSegments() {
  ...
  newIndexSegments = new ConcurrentSkipListMap<>(activeIndexSegments);
  newIndexSegments.putAll(segmentsToAdd);
  newIndexSegments.keySet().removeAll(segmentsToRemove);
  activeIndexSegments = newIndexSegments;
  newIndexSegments = null;
  indexesRef.set(activeIndexSegments);
}

addToIndex() {
  ...
  activeIndexSegments.put(key);
  if (newIndexSegments != null) {
    newIndexSegments.put(key); // put in both old and new.
  }
  journal.addEntry();
}
```
The above is likely not without bugs - but I hope it conveys what I am trying to say. You can figure out the right way and tests can validate it. This should not make the code all that ugly as I originally feared, and is worth the benefit of getting rid of the lock.

vgkholla replied:
I'm not sure I follow why this is thread safe. The statement `newIndexSegments = new ConcurrentSkipListMap<>(activeIndexSegments)` is not atomic (i.e copying the existing map and setting/updating the reference of `newIndexSegments`). So you can still have a situation where the copy is finished but the `newIndexSegments` variable has not been set.  In that time, if `addToIndex()` starts and finishes executing, you would still have missed out on the new key

Other than that, I'm not sure why we need to complicate this further - what the expected gain out of removing that lock if it is infrequently (almost never) held and simplifies the code?

vgkholla replied:
Also, I do think the lock is inescapable. The lock is not on the data structure itself but on the reference change. If the same reference were being used, operating on different regions would not require a lock.

pnarayanan replied:
Just to recap what we discussed and concluded in person. 
- The lock is likely avoidable. If the code gets too complicated, we could go with the lock as even though `addToIndex()` is in the critical path, it is called by the store after taking a store lock, so the contention for this lock will not be an issue. If not, we should get rid of the lock as we rarely use locks within the index (if at all) and because we use concurrent data structures and atomic references everywhere and generally follow the lock-free paradigm within the index.
- `ConcurrentSkipListMap` provides weakly consistent iterators which can be used to get a snapshot view of the old map at the point of creation or since creating it. What will need to be ensured is that the happened-before relation between the start of this iteration and `addToIndex()` adding to both old and new references, is maintained.
- We also discussed whether we would really need atomic references. The original reason was to save off the value at the start of methods that need to look at the segments map. This may be possible even without atomic references.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -333,9 +390,14 @@ void addToIndex(IndexEntry entry, FileSpan fileSpan) throws StoreException {
           entry.getValue().getBytes().capacity(), config, metrics);
       info.addEntry(entry, fileSpan.getEndOffset());
-      indexes.put(info.getStartOffset(), info);
+      indexesChangeLock.lock();
```
From what I see, `changeIndexSegments()` will never get called to alter the trailing part of the Index (covered by the `Journal`), and it has explicit checks to guard against that. For the same reason that you do not have the `journal.addEntry()` covered by this lock here, I think the index put doesn't need a lock either. You should be able to deference the atomic reference and simply do the put. Am I missing something?

Also, I don't see the need for the lock in general (see my other comment), but maybe you have other reasons for that which are not part of this patch.

vgkholla replied:
Discussion is the same as [this](https://github.com/linkedin/ambry/pull/543#discussion_r102541125).

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -96,6 +97,11 @@ public int compare(File o1, File o2) {
 
+  // switching this is thread safe as long as there are no modifications to IndexSegment instances whose offsets are
+  // still present in the journal.
+  private final AtomicReference<ConcurrentSkipListMap<Offset, IndexSegment>> indexesRef =
```
Done - changed similar names everywhere in the class.

vgkholla commented:
The major changes in the previous commit are
1. Got rid of the `AtomicReference` since it was not adding value
2. Based on the [comment and suggestions](https://github.com/linkedin/ambry/pull/543#discussion_r102541125) from @pnarayanan, have removed the lock in `PersistentIndex`
3. Added a test to verify that there are no concurrency problems b/w `addToIndex()` and `changeIndexSegments()`.

Thanks @pnarayanan for the great suggestions/help on making the whole thing lock free !

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -523,13 +584,17 @@ BlobReadOptions getBlobReadInfo(StoreKey id, EnumSet<StoreGetOptions> getOptions
+              indexSegments);
       if (value.getOriginalMessageOffset() != -1) {
         // PUT record in the same log segment.
         String logSegmentName = value.getOffset().getName();
```
What we concluded from the offline discussion was that this isn't necessary. We have no problems with correctness and the guarantees provided by this function are not changing.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -625,16 +690,19 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
+        ConcurrentSkipListMap<Offset, IndexSegment> indexSegments = validIndexSegments;
 
-        if (entries != null) {
+        if (entries != null && journal.getFirstOffset().compareTo(offsetToStart) <= 0) {
```
Removed the check

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -625,16 +690,19 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
           for (JournalEntry entry : entries) {
-            IndexValue value = findKey(entry.getKey());
+            IndexValue value =
+                findKey(entry.getKey(), new FileSpan(entry.getOffset(), getCurrentEndOffset(indexSegments)),
```
That's correct. As suggested, I have called this out via a comment so that it is clear when reading code.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -854,31 +946,37 @@ private StoreFindToken findEntriesFromSegmentStartOffset(Offset initialSegmentSt
           }
           newTokenOffsetInJournal = entry.getOffset();
-          IndexValue value = findKey(entry.getKey());
+          IndexValue value =
```
We discussed this offline - recording it here.

This situation has been taken care of. The code that precedes this has been added to tackle exactly this situation
```
// stop if there are no more index segments in this ref.
if (nextSegmentStartOffset == null) {
  break;
}
```
Whenever a journal entry that does not exist is encountered, it will have caused a rollover of the index segment. If that index segment does not exist in the world-view that we have, we break out. I have added a comment to call this out explicitly.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1081,22 +1203,25 @@ FindInfo findDeletedEntriesSince(FindToken token, long maxTotalSizeOfEntries, lo
 
         Offset offsetEnd = offsetToStart;
-        if (entries != null) {
+        if (entries != null && journal.getFirstOffset().compareTo(offsetToStart) <= 0) {
```
Removed.

vgkholla commented:
@pnarayanan Addressed your comments and double checked that it is necessary to save off references where it is done and that snapshots are either taken after fetching entries from the journal or care is taken not to make an assumption that all entries returned by the journal are present in the index.

vgkholla commented:
Formatted and built on both OSx and Linux. The tests for concurrency correctness b/w `addToIndex()` and `changeIndexSegments()` runs 500 times and that should account for any bugs there too.
